### DISCUSSION TITLE
Why does `write` function return `int` (signed 32 bit), but take `size_t` (unsigned 64 bit) parameter?

### DISCUSSION THREAD

**Discussion by xBZZZZ** [libfuse/libfuse#791] [Q&A]
https://github.com/libfuse/libfuse/blob/4c177c904778c45ebd992b7f651af4f0039c550c/include/fuse.h#L477
https://github.com/libfuse/libfuse/blob/4c177c904778c45ebd992b7f651af4f0039c550c/include/fuse.h#L484
Why __the number of bytes requested__ (return `int` (signed 32 bit)) has smaller type than __the number of bytes requested__ (parameter `size_t` (unsigned 64 bit))?

**Comment by trapexit** [comment]
Plausibility from when things were 32bit. Or because the max write is 2GiB on Linux. Or because it was just missed.

**Comment by Nikratio** [comment]
The callback needs to be able to signal when there's an error. But you are right that `int` is not a good choice, `ssize_t` exists for exactly this purpose.

**Comment by bsbernd** [comment]
Probably missed, but then also fuse does not support more than 1M writes right now (before 128kb) - unlikely to ever hit the int limit.
